protobuf/lib/src/protobuf/field_set.dart

```diff
@@ -792,14 +799,15 @@ class _FieldSet {
     }
 
     if (otherFi.isGroupOrMessage) {
-      final currentFi = isExtension!
+      GeneratedMessage? currentFi = isExtension!
```

Comment:
You could consider doing the `as GeneratedMessage?` check only in the non-null case.